File: patternlets/MPI/11.broadcast2/broadcast2.c
/* broadcast2.c
 * ... illustrates the use of MPI_Bcast() for arrays...
 * Joel Adams, Calvin College, November 2009.
 *
 * Usage: mpirun -np N ./broadcast2
 *
 * Exercise:
 * - Compile and run, using 2, 4, and 8 processes
 * - Use source code to trace execution and output
 * - Explain behavior/effect of MPI_Bcast().
 */

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

/* fill an array with some arbitrary values
 * @param: a, an int*.
 * @param: size, an int.
 * Precondition: a is the address of an array of ints.
 *              && size is the number of ints a can hold.
 * Postcondition: a has been filled with arbitrary values
 *                { 11, 12, 13, ... }.
 */
void fill(int* a, int size) {
	int i;
	for (i = 0; i < size; i++) {
		a[i] = i+11;
	}
}

/* display a string, a process id, and its array values
 * @param: str, a char*
 * @param: id, an int
 * @param: a, an int*.
 * Precondition: str points to either "BEFORE" or "AFTER"
 *              && id is the rank of this MPI process
 *              && a is the address of an 8-element int array.
 * Postcondition: str, id, and a have all been written to stdout.
 */
void print(char* str, int id, int* a) {
	printf("%s broadcast, process %d has: {%d, %d, %d, %d, %d, %d, %d, %d}\n",
	   str, id, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]);
}

#define MAX 8

int main(int argc, char** argv) {
	int array[MAX] = {0};
	int numProcs, myRank;

	MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &numProcs);
	MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

	if (myRank == 0) fill(array, MAX);

	print("BEFORE", myRank, array);

	MPI_Bcast(array, MAX, MPI_INT, 0, MPI_COMM_WORLD);

	print("AFTER", myRank, array);

	MPI_Finalize();

	return 0;
}

File: patternlets/java-OpenMPI/12.broadcast2/arrayVersion/Broadcast2.java
/* Broadcast2.java
 * ... illustrates the use of MPI's broadcast command with multiple values.
 *
 * Note: This version uses an array to store the values.
 *
 * Goal: The master process fills an array with values
 *        and broadcasts it to all the other processes.
 *       Each process outputs its array before and after
 *        the broadcast.
 *
 * Joel Adams, Calvin University, November 2019,
 *
 * Usage: mpirun -np 4 java ./Broadcast2
 *
 * Exercise:
 * - Compile, then run several times,
 *     using 2, 4, and 8 processes
 * - Use source code to trace execution and output
 * - Explain behavior/effect of MPI_Bcast().
 */

import mpi.*;

public class Broadcast2 {

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    Comm comm         = MPI.COMM_WORLD;
    int id            = comm.getRank();

    int [] array = new int[ARRAY_SIZE];            // all: allocate array 

    if ( id == MASTER ) {                          // MASTER: fill its array
        fill(array);
    }

    print("BEFORE", id, array);                    // all: print buffers before

    printSeparator("----", id, comm);

    comm.bcast(array, array.length, MPI.INT, 0);   // all: participate in broadcast

    print("AFTER", id, array);                     // all: print buffers after

    MPI.Finalize();
  }


  /* utility to fill an array with some values.
   * @param: a, an int array.
   * POST: a has been filled with int values.
   */
  private static void fill(int [] a) {
     for (int i = 0; i < a.length; ++i) {
         a[i] = i + 11;
     }
  }

  /* utility to print a buffer with descriptive labels.
   * @param: label, a String.
   * @param: id, this process's MPI rank.
   * @param, a, an int array.
   * POST: label, id, and a have been displayed via System.out.
   */
  private static void print(String label, int id, int [] a) {
    String msg = label + " the broadcast, process " + id
                       + "'s array contains:";
    for (int i = 0; i < a.length; ++i) {
      msg += (" " + a[i]);
    }
    msg += "\n";
    System.out.print(msg); 
  }
 
  /* utility to print a separator string between the 'before' and 'after' parts.
   * @param: separator, a String.
   * @param: id, the rank of this MPI process.
   * @param: comm, the Communicator for the processes involved.
   * POST: the master has printed the separator to System.out.
   */
  public static void printSeparator(String separator, int id, Comm comm) 
                                      throws MPIException {
     comm.barrier();
     if (id == MASTER) { System.out.println(separator); }
     comm.barrier();
  }

  private static final int MASTER = 0;
  private static final int ARRAY_SIZE = 8;
}

